server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4307,6 +4307,35 @@ public void afterRefresh(boolean didRefresh) {
         }
     }
 
+    /**
+     * Reset the current engine to a new one.
+     *
+     * Calls {@link Engine#prepareForEngineReset()} on the current engine, then closes it, and loads a new engine without
+     * doing any translog recovery.
+     *
+     * In general, resetting the engine should be done with care, to consider any in-progress operations and listeners.
+     * At the moment, this is implemented in serverless for a special case that ensures the engine is prepared for reset.
+     */
+    public void resetEngine() {
+        assert Thread.holdsLock(mutex) == false : "resetting engine under mutex";
+        assert waitForEngineOrClosedShardListeners.isDone();
+        try {
+            final var currentEngine = getEngine();
+            currentEngine.prepareForEngineReset();
```

Comment:
I wonder if this should be called with the engineMutex held?

Reply:
Good idea, done.